src/main/java/bookstore/mapper/BookMapper.java

```diff
@@ -2,13 +2,32 @@
 
 import bookstore.config.MapperConfig;
 import bookstore.dto.book.BookDto;
+import bookstore.dto.book.BookWithoutCategoryIds;
 import bookstore.dto.book.CreateBookRequestDto;
 import bookstore.model.Book;
+import bookstore.model.Category;
+import java.util.stream.Collectors;
+import org.mapstruct.AfterMapping;
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
 
 @Mapper(config = MapperConfig.class)
 public interface BookMapper {
+
+    BookWithoutCategoryIds toBookWithoutCategoryId(Book book);
+
+    @Mapping(target = "categoryIds", ignore = true)
     BookDto toDto(Book book);
 
     Book toModel(CreateBookRequestDto requestDto);
+
+    @AfterMapping
+    default void setCategoryIds(@MappingTarget BookDto bookDto, Book book) {
+        if (book.getCategories() != null) {
```

Comment:
It's never really null, as we have initialized it like that: Set<Category> categories = new HashSet<>();
you may use !book.getCategories().isEmpty() instead